src/transport_picoquic.cpp

```diff
@@ -0,0 +1,828 @@
+  stream_cnx->stream_id = stream_id;
+  stream_cnx->context_id = (uint64_t)cnx;
+  stream_cnx->cnx = cnx;
+  stream_cnx->in_data.setLimit(1000);
+  stream_cnx->out_data.setLimit(1000);
+
+
+  sockaddr* addr;
+  picoquic_get_peer_addr(cnx, &addr);
+  std::memset(stream_cnx->peer_addr_text, 0, sizeof(stream_cnx->peer_addr_text));
+
+  switch (addr->sa_family) {
+    case AF_INET:
+      (void)inet_ntop(AF_INET,
+                      (const void*)(&((struct sockaddr_in*)addr)->sin_addr),
+                      stream_cnx->peer_addr_text,
+                      sizeof(stream_cnx->peer_addr_text));
+      stream_cnx->peer_port = ntohs(((struct sockaddr_in*) addr)->sin_port);
+      break;
+
+    case AF_INET6:
+      (void)inet_ntop(AF_INET6,
+                      (const void*)(&((struct sockaddr_in6*)addr)->sin6_addr),
+                      stream_cnx->peer_addr_text, sizeof(stream_cnx->peer_addr_text));
+      stream_cnx->peer_port = ntohs(((struct sockaddr_in6*) addr)->sin6_port);
+      break;
+  }
+
+  (void)picoquic_set_app_stream_ctx(cnx, stream_id, stream_cnx);
+
+  if (stream_id)
+      (void)picoquic_mark_active_stream(cnx, stream_id, 1, stream_cnx);
+
+  return stream_cnx;
+}
+
+PicoQuicTransport::PicoQuicTransport(const TransportRemote& server,
+                                     const TransportConfig& tcfg,
+                                     TransportDelegate& delegate,
+                                     bool isServerMode,
+                                     LogHandler& logger)
+    : logger(logger)
+      , isServerMode(isServerMode)
+      , stop(false)
+      , transportStatus(TransportStatus::Connecting)
+      , serverInfo(server)
+      , delegate(delegate)
+{
+  if (isServerMode && tcfg.tls_cert_filename == NULL) {
+    throw InvalidConfigException("Missing cert filename");
+  }
+  else if (tcfg.tls_cert_filename != NULL) {
+    (void)picoquic_config_set_option(&config, picoquic_option_CERT,
+                               tcfg.tls_cert_filename);
+
+    if (tcfg.tls_key_filename != NULL) {
+      (void)picoquic_config_set_option(&config, picoquic_option_KEY,
+                                 tcfg.tls_key_filename);
+    } else {
+      throw InvalidConfigException("Missing cert key filename");
+    }
+  }
+}
+
+PicoQuicTransport::~PicoQuicTransport()
+{
+  setStatus(TransportStatus::Shutdown);
+  shutdown();
+}
+
+void PicoQuicTransport::shutdown()
+{
+
+  if (stop) // Already stopped
+    return;
+
+  stop = true;
+
+  if (picoQuicThread.joinable()) {
+    logger.log(LogLevel::info, "Closing transport pico thread");
+    picoQuicThread.join();
+  }
+
+  cbNotifyQueue.stopWaiting();
+  if (cbNotifyThread.joinable()) {
+    logger.log(LogLevel::info, "Closing transport callback notifier thread");
+    cbNotifyThread.join();
+  }
+
+  logger.log(LogLevel::info, "done closing transport threads");
+
+}
+
+
+TransportStatus
+PicoQuicTransport::status()
```

Comment:
Should this be a const member function?

Reply:
That should be fine. I'll update it. It does require updates to the interface since it's an implemented method.